### PR TITLE
[SPARK-28770][CORE][TEST] Fix ReplayListenerSuite tests that sometimes fail

### PR DESCRIPTION
**Pull request by wypoon**
### What changes were proposed in this pull request?

`ReplayListenerSuite` depends on a listener class to listen for replayed events. This class was implemented by extending `EventLoggingListener`. `EventLoggingListener` does not log executor metrics update events, but uses them to update internal state; on a stage completion event, it then logs stage executor metrics events using this internal state. As executor metrics update events do not get written to the event log, they do not get replayed. The internal state of the replay listener can therefore be different from the original listener, leading to different stage completion events being logged.

We reimplement the replay listener to simply buffer each and every event it receives. This makes it a simpler yet better tool for verifying the events that get sent through the ReplayListenerBus.

### Why are the changes needed?

As explained above. Tests sometimes fail due to events being received by the `EventLoggingListener` that do not get logged (and thus do not get replayed) but influence other events that get logged.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Existing unit tests.

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala [hunk 1]
```diff
@@ -219,11 +219,25 @@ class ReplayListenerSuite extends SparkFunSuite with BeforeAndAfter with LocalSp
...
+    // Stage executor metrics are logged at stage completion, for any of the executors and
+    // the driver for which we have metrics. We always have metrics for executors, because
+    // they are sent at task end as well as in executor metrics updates. We do not always
+    // have metrics for the driver, because they are only sent in executor metrics updates
+    // (at heartbeat intervals), and when we do, it is only in the original events, never
+    // in the replay, since executor metrics updates are not replayed.
+    // For this reason, exclude stage executor metrics for the driver.
+    val filteredEvents = originalEvents.filter { e =>
+      if (e.isInstanceOf[SparkListenerStageExecutorMetrics] &&
+        e.asInstanceOf[SparkListenerStageExecutorMetrics].execId == "driver") {
```
Comment (HyukjinKwon) [inline core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala]: What I am not understanding is that, why are both different? They make JSONs from the same instances. Are they different? Why is it flaky?

File: core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala [hunk 2]
```diff
@@ -219,11 +219,25 @@ class ReplayListenerSuite extends SparkFunSuite with BeforeAndAfter with LocalSp
...
+      .map(JsonProtocol.sparkEventFromJson(_))
+    // Executor metrics updates are not logged, so do not get replayed.
+    // Stage executor metrics are logged at stage completion, for any of the executors and
+    // the driver for which we have metrics. We always have metrics for executors, because
+    // they are sent at task end as well as in executor metrics updates. We do not always
+    // have metrics for the driver, because they are only sent in executor metrics updates
+    // (at heartbeat intervals), and when we do, it is only in the original events, never
+    // in the replay, since executor metrics updates are not replayed.
+    // For this reason, exclude stage executor metrics for the driver.
+    val filteredEvents = originalEvents.filter { e =>
```
Comment (HeartSaVioR) [inline core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala]: I guess below would be simpler:

```
val filteredEvents = originalEvents.filter {
  case e: SparkListenerStageExecutorMetrics if e.execId == "driver" => false
  case _ => true
}
```

**Comment by HeartSaVioR** [comment]
I'd rather say we may need to find the way to let EventMonster not extending EventLoggingListener and just buffering all events. We don't need to deal with such complicated case in EventLoggingListener, as what we want to verify in ReplayListenerSuite is "ReplayListener can read all events sequentially and pass to registered listeners correctly".

https://github.com/apache/spark/blob/df39855db826fd4bead85a2ca01eda15c101bbbe/core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala#L236-L253

I'm a bit concerned about the javadoc of EventMonster - I expect each event are posted to be same to all listeners. If that's the understanding of us (Spark community), it would end up with inconsistency if we allow the event itself or any fields in event to be modified somewhere (e.g. #25672).

File: core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala (hunk 1 above)
Comment (HeartSaVioR) [reply core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala]: EventMonster extends EventLoggingListener which doesn't transparently log events, hence some possible cases could exist where both can be different.

Comment (wypoon) [reply core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala]: I tried to explain briefly in the comment. Let me try to explain in more detail.

When the application is run, executor metrics update events may occur (if the application runs long enough). These events are not written to the event log, so they do not get replayed. That is the root cause of differences.
On executor metrics update, the EventLoggingListener updates a map it uses to track per-stage metrics, but does not log. The listener also receives metrics on task end. Because of this, it always has metrics for executors. On stage completion. the listener logs stage executor metrics, for any of the executors/driver it has metrics for. As explained, sometimes it will have metrics for the driver (if an executor metrics update arrived from the driver), but most of the time not.
To recap, SparkListenerStageExecutorMetrics events are not events received while the application is running, but secondary events derived from metrics received and written to the event log. (One other point not relevant to this bug but may help understanding - on replay, the SparkListenerStageExecutorMetrics events do get replayed but are ignored by the listener.)

File: core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala (hunk 2 above)
Comment (wypoon) [reply core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala]: Yes, that's nicer. Thanks.

**Comment by wypoon** [comment]
> I'd rather say we may need to find the way to let EventMonster not extending EventLoggingListener and just buffering all events. We don't need to deal with such complicated case in EventLoggingListener, as the purpose of ReplayListenerSuite is to test the fact that "ReplayListener can read all events sequentially and pass to registered listeners correctly".
> 
I'd rather not take on reworking EventMonster here. I think it is sufficient to handle the special case of `SparkListenerStageExecutorMetrics` by filtering.

File: core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala (hunk 1 above)
Comment (wypoon) [reply core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala]: I had written the same explanation in https://github.com/apache/spark/pull/25659#issuecomment-527741025.

**Comment by HeartSaVioR** [comment]
The test is going to be unnecessary complicated, as I commented earlier. We are required to know about detailed behavior of EventLoggingListener, though we are in ReplayListenerSuite. ReplayListener is even not used with EventLoggingListener so it's really only for testing which doesn't make sense to couple with. Even we fix it for this time, the test might be broken again if we tune EventLoggingListener once more.

For sure, it's not your fault. You did a great analysis. The code looks to be ancient, so it's not their fault too. We just seem to miss the relevant codes while EventLoggingListener is evolving. But at least now we indicate the unnecessary complexity, it's your bet to just add band-aid or fix it on the right way. I prefer latter, and I'm OK if committers are OK with former and merge the patch as it is. Maybe I'll raise another PR to fix it then.

**Comment by wypoon** [comment]
@HeartSaVioR, on second thoughts, after offline discussion with @squito, I agree that the best way to fix the tests is to reimplement EventMonster. Instead of extending EventLoggingListener, EventMonster should simply buffer any and all events it receives. @squito pointed out that EventMonster was added when ReplayListenerSuite was first added, and that SparkFirehoseListener did not exist then. Extending EventLoggingListener and adding a mode where it simply appends events to a buffer (in its logEvent method) was an easy way of implementing EventMonster. However, as EventLoggingListener evolved, it did not call logEvent on every event, and it also wrote events as side effects that depended on its internal state, leading to the current difficulties.
I have reimplemented EventMonster (extending SparkFirehoseListener) to simply buffer each and every event it receives. This matches the needs of ReplayListenerSuite better. With this change, testApplicationReplay no longer needs any logic to filter any events.

**Review by HeartSaVioR — APPROVED** [review]
LGTM

**Comment by HeartSaVioR** [comment]
cc. @squito as he commented and discussed the approach of resolution in JIRA issue.

File: core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala [hunk 3]
```diff
@@ -236,20 +238,15 @@ class ReplayListenerSuite extends SparkFunSuite with BeforeAndAfter with LocalSp
...
-   * processed by one listener and being processed by another. Thus, in order to establish
-   * a fair comparison between the original events and the replayed events, both functionalities
-   * must be implemented within one listener (i.e. the EventLoggingListener).
-   *
-   * This child listener inherits only the event buffering functionality, but does not actually
-   * log the events.
    */
-  private class EventMonster(conf: SparkConf)
-    extends EventLoggingListener("test", None, new URI("testdir"), conf) {
+  private class EventMonster extends SparkFirehoseListener {
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala]: No need to change it here, but, Event_Monster_?

Comment (squito) [reply core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala]: lol, good point.  EventBufferingListener would be better

**Comment by squito** [comment]
lgtm.  Also I see @bzhaoopenstack confirmed this does the trick on https://github.com/apache/spark/pull/25659.

wing yew, can you do the rename?  and update the PR descrpition to not mention EventMonster by name anymore

File: core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala (hunk 3 above)
Comment (wypoon) [reply core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala]: Sure, I will rename it EventBufferingListener.

Comment (wypoon) [reply core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala]: Updated the PR description as well to not mention EventMonster.

**Comment by squito** [comment]
merged to master, thanks everyone
